src/common/models/data-series/data-series-names.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2017-2018 Allegro.pl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SeriesDerivation } from "../series/series";
+import { DataSeriesPercentOf } from "./data-series";
+
+function formatPercentOf(percentOf: DataSeriesPercentOf): string {
+  return `__${percentOf}_`;
+}
+
+function formatDerivation(derivation: SeriesDerivation): string {
+  return `_${derivation}__`;
+}
+
+export function fullName(name: string, derivation: SeriesDerivation, percentOf?: DataSeriesPercentOf): string {
+  const percentStr = percentOf ? formatPercentOf(percentOf) : "";
+  const derivationStr = derivation === SeriesDerivation.CURRENT ? "" : formatDerivation(derivation);
+  return `${derivationStr}${name}${percentStr}`;
+}
+
+function extractDerivation(fullName: string): { derivation: SeriesDerivation, length: number } {
+  if (fullName.startsWith(formatDerivation(SeriesDerivation.PREVIOUS))) {
+    return { derivation: SeriesDerivation.PREVIOUS, length: formatDerivation(SeriesDerivation.PREVIOUS).length };
+  }
+  if (fullName.startsWith(formatDerivation(SeriesDerivation.DELTA))) {
+    return { derivation: SeriesDerivation.DELTA, length: formatDerivation(SeriesDerivation.DELTA).length };
+  }
+  return { derivation: SeriesDerivation.CURRENT, length: 0 };
+}
+
+function extractPercent(fullName: string): { percentOf?: DataSeriesPercentOf, length: number } {
+  if (fullName.endsWith(formatPercentOf(DataSeriesPercentOf.TOTAL))) {
+    return { percentOf: DataSeriesPercentOf.TOTAL, length: formatPercentOf(DataSeriesPercentOf.TOTAL).length };
+  }
+  if (fullName.endsWith(formatPercentOf(DataSeriesPercentOf.PARENT))) {
+    return { percentOf: DataSeriesPercentOf.PARENT, length: formatPercentOf(DataSeriesPercentOf.PARENT).length };
+  }
+  return { length: 0 };
+}
+
+export function nominalName(fullName: string): { name: string, derivation: SeriesDerivation, percentOf?: DataSeriesPercentOf } {
```

Comment:
When we generate plywood query, we need a name for each expression. So for example for "Clicks" we can use measure name, but "Previous Clicks" or "Clicks (% of total)" we need serialize DataSeries to string (with small plywood caveats). In two places, we need to invert that process:

1. When generating TSV file, we get plywood Datum, which have those names as keys. TSV converter in plywood reads those keys and passes to our code so we can change it to readable title. So we need to decipher something like "_previous__clicks__of_parent_" and change it to "Previous Clicks (% of Parent)". I don't see a better way for now.

2. When sorting data in Table viz, we also need to construct name for expression and later read it back. This could be changed though. As you said, we could change SortOn type to keep some union of DataSeries and Dimension and serialize it lazily.

Reply:
For first case I could iterate through all `DataSeries` in essence and compare fullName with string from plywood.